### PR TITLE
feat: add stage cache foundation

### PR DESCRIPTION
**Pull request by lox**
## Summary
- add a dedicated `cachestore` for system-managed stage cache metadata
- add canonical stage key builders for runtime, workspace, and dependency cache stages
- extract a shared stored-rootfs restore path so user snapshots and stage caches can restore without sharing metadata stores
- migrate workspace-stage lookup and publish flow off reserved snapshot names and onto `cachestore`
- keep `snapshotstore` user-facing only and update tests accordingly

## Validation
- `mise exec -- go test ./internal/cachekey ./internal/cachestore ./internal/controlservice ./internal/cli`
- `mise exec -- go test ./...`

## Follow-up
- dependency-stage publication and policy surface are still a separate slice
- cache GC and cleanup of replaced backing snapshots remain follow-up work

### CODE REVIEW

File: internal/controlservice/workspace_seed.go [hunk 1]
```diff
@@ -2,91 +2,89 @@ package controlservice
...
+	if !ok {
+		return cachestore.Record{}, false, nil
 	}
-	record, ok := workspaceSeedSnapshotRecord(records, backendName, compiled.Hash, runtimeBaseKey, repository)
-	return record, ok, nil
+	if strings.TrimSpace(record.Backend) != strings.TrimSpace(backendName) {
+		return cachestore.Record{}, false, nil
+	}
+	if !repositoryCheckoutsEqual(repositorycheckout.FromProto(record.Repository), repository) {
+		return cachestore.Record{}, false, nil
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/controlservice/workspace_seed.go:88]: Fixed in 5b55b13: workspace-stage cache keys now include the compiled policy hash, and lookup still rejects any record whose stored policy hash does not match the current compiled policy.

**Comment by lox** [comment]
@codex review

File: internal/controlservice/service.go [hunk 2]
```diff
@@ -308,100 +323,31 @@ func (s *Service) createSandboxFromSnapshot(ctx context.Context, req *cleanroomv
...
+	snapshotID := strings.TrimSpace(record.SnapshotID)
+	if snapshotID == "" {
+		return nil, errors.New("missing snapshot_id")
 	}
-	if !snapshotOperationsEnabledForBackend(backendName, s.Config) {
-		return nil, fmt.Errorf("snapshots are not enabled for backend %q", backendName)
+	if err := s.beginSnapshotUse(snapshotID); err != nil {
+		return nil, err
 	}
+	defer s.finishSnapshotUse(snapshotID)
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/controlservice/service.go]: Fixed in b21baf7: `createSandboxFromSnapshot` now acquires the snapshot-use lock before loading snapshot metadata, and the new test blocks `store.Get` to verify delete is rejected while that load is in flight.

**Comment by lox** [comment]
@codex review

File: internal/controlservice/workspace_seed.go [hunk 3]
```diff
@@ -2,91 +2,93 @@ package controlservice
...
-	return strings.HasPrefix(strings.TrimSpace(name), workspaceSeedSnapshotNamePrefix)
+func workspaceStageCacheKey(runtimeBaseKey, compiledPolicyHash string, repository *repositorycheckout.Checkout) string {
+	if repository == nil || strings.TrimSpace(runtimeBaseKey) == "" || strings.TrimSpace(compiledPolicyHash) == "" {
+		return ""
+	}
+	return cachekey.WorkspaceStageKey(cachekey.WorkspaceStageInputs{
+		RuntimeKey:                  strings.TrimSpace(runtimeBaseKey),
+		CompiledPolicyHash:          strings.TrimSpace(compiledPolicyHash),
+		CanonicalRemoteURL:          strings.TrimSpace(repository.RemoteURL),
+		CommitSHA:                   strings.TrimSpace(repository.CommitSHA),
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/controlservice/workspace_seed.go]: Fixed in 8fff7de: workspace-stage cache keys now include backend identity, and the new regression test proves firecracker and darwin-vz can publish separate workspace-stage cache entries without duplicate-key rollbacks.

**Comment by lox** [comment]
@codex review

File: internal/controlservice/workspace_seed.go [hunk 4]
```diff
@@ -2,91 +2,94 @@ package controlservice
...
+	}
+	if !ok {
+		return cachestore.Record{}, false, nil
+	}
+	if strings.TrimSpace(record.Backend) != strings.TrimSpace(backendName) {
+		return cachestore.Record{}, false, nil
 	}
-	record, ok := workspaceSeedSnapshotRecord(records, backendName, compiled.Hash, runtimeBaseKey, repository)
-	return record, ok, nil
+	if !repositoryCheckoutsEqual(repositorycheckout.FromProto(record.Repository), repository) {
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/controlservice/workspace_seed.go:87]: Fixed in f3f803d: workspace-stage lookup now normalizes checkout fields before comparing metadata, and the new regression test proves `/workspace/` and `/workspace` reuse the same cache entry instead of cold-bootstrapping and rolling back a duplicate publish.

**Comment by lox** [comment]
@codex review

File: internal/cli/serve.go [hunk 5]
```diff
@@ -202,12 +204,17 @@ func newControlService(ctx *runtimeContext, logger *log.Logger, mirrors gateway.
 	if err != nil {
 		return nil, fmt.Errorf("configure snapshot metadata store: %w", err)
 	}
+	cacheMetadataStore, err := newCacheMetadataStore(cachestore.Options{})
+	if err != nil {
+		return nil, fmt.Errorf("configure cache metadata store: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/cli/serve.go]: Fixed in c82ac17: cache metadata store init is now best-effort in `newControlService`, so `cleanroom serve` still starts with stage caches disabled when cache metadata setup fails, and the new CLI test covers that degraded startup path.

File: internal/controlservice/workspace_seed.go [hunk 6]
```diff
@@ -128,34 +137,71 @@ func (s *Service) maybePublishWorkspaceSeedSnapshot(
...
 			return
 		}
-		s.logWorkspaceSeedWarning("persist workspace seed snapshot metadata", sandboxID, err)
+		s.logWorkspaceSeedWarning("persist workspace stage cache metadata", sandboxID, err)
+		return
+	}
+
+	if replacedRecord != nil && strings.TrimSpace(replacedRecord.CacheKey) == cacheKey {
+		if err := deleteWorkspaceStageCacheSnapshot(ctx, adapter, backendName, firecrackerCfg, *replacedRecord); err != nil {
+			s.logWorkspaceSeedWarning("delete replaced workspace stage cache snapshot", sandboxID, err)
```
[the remark this answers is not in the document]
Comment (lox) [reply internal/controlservice/workspace_seed.go:182]: Fixed in c82ac17: stage-cache restores now register backing snapshot use before provisioning, and replaced-cache cleanup takes the same snapshot delete lock as user snapshots. The new regression test proves cleanup is rejected while a warm cache restore is in flight.

**Comment by lox** [comment]
@codex review
